archivesunleashed/aut
# make ArchiveRecord a trait

**GitHub issue(s)**:

* https://github.com/archivesunleashed/aut/pull/175
* https://github.com/archivesunleashed/aut/pull/181

# What does this Pull Request do?

This commit introduces a trait that describes the interface of an `ArchiveRecord` in AUT. This allows for alternative implementations of this core record type in AUT, such as the one provided by [ArchiveSpark-AUT-bridge](https://github.com/helgeho/ArchiveSpark-AUT-bridge). The original AUT implementation of `ArchiveRecord`, which is instantiated by AUT's `RecordLoader`, is moved to the `ArchiveRecordImpl` class.

This was already done before by https://github.com/archivesunleashed/aut/pull/175 and reverted by https://github.com/archivesunleashed/aut/pull/181 because the change of moving `ISO8601` into a *static object* caused some issues with certain collections (see below). Therefore, this time, `ISO8601` remains part of `ArchiveRecordImpl`.

# How should this be tested?

As this is a pure refactoring without changing any logic, a simple compilation + smoke test should be sufficient. However, as it turns out, the little additional change of moving `ISO8601` into a *static object* in the first attempt of this change (https://github.com/archivesunleashed/aut/pull/175) caused strange behaviors with some collections (probably when running on multiple cores per executor, since `SimpleDateFormat` is not thread-safe as pointed out by @anjackson, see https://github.com/archivesunleashed/aut/pull/181). Hence, it should be better tested with different kinds of collections to make sure there are no unexpected side effects involved.

# Additional Notes:

The first attempt to make ArchiveRecord a trait (https://github.com/archivesunleashed/aut/pull/175) was reverted by https://github.com/archivesunleashed/aut/pull/181. This time nothing else is changed except for the trait (`ISO8601` remains part of `ArchiveRecordImpl`). @ruebot and @ianmilligan1 please test it as you did last time (https://github.com/archivesunleashed/aut/pull/181), I hope it is not causing problems this time.
